doc/developer/design/20241028_optimizer_releases.md

```diff
@@ -0,0 +1,81 @@
+  - targeting new dataflow operators
+  - changing AST types for HIR, MIR, or LIR
+
+Optimizer engineers will be able to deploy hotfixes to cloud customers.
+
+## Solution Proposal
+
+We propose the following solution:
+
+  1. Refactor existing `sql`, `expr`, `compute-types`, and `transform` crates into `optimizer` and `optimizer-types` crates.
+  2. Version the `optimizer` crate. The `optimizer-types` crate will contain traits and types that are stable across all three versions (e.g., sufficient interfaces to calculate the dependencies of various expression `enum`s without fixing any of their details).
+  3. Work on three versions of the `optimizer` crate at once:
+     - **unstable**/**hot** active development
+     - **testing**/**warm** most recently cut release of the optimizer
+     - **stable**/**cold** previously cut release of the optimizer
+     These versions will be tracked as copies of the crate in the repo (but see ["Alternatives"](#alternatives)).
+  4. Rotate versions treadmill style, in some way matching the cadence of the self-hosted release. (Likely, we will want to use the cloud release to qualify a self-hosted release.)
+
+## Minimal Viable Prototype
+
+### Version names and numbering
+
+Every version of the `optimizer` crate will be numbered. Customers can run SQL like:
+
+```sql
+ALTER CLUSTER name SET (OPTIMIZER VERSION = version)
+```
+
+where `version` is a version number (or, possibly one of our three fixed names). Note that the optimizer is selected on a per cluster basis. Selecting the optimizer per cluster makes it easier to qualify optimizer releases.
+
+### The `optimizer` and `optimizer-types` crates
+
+The `optimizer` create will containe the definitions of HIR, MIR, and LIR, along with the HIR-to-MIR and MIR-to_LIR lowerings and the MIR-to-MIR transformations. These come from the `sql` (HIR, HIR -> MIR), `expr` (MIR), `transform` (MIR -> MIR), and `compute-types` (LIR, MIR -> LIR) crates.
+
+The `optimizer-types` crate will have traits and definitions that are global across all three live versions of the optimizer. The AST types may change version to version, but the traits will be more stable.
+
+These crates do _not_ include:
+  + the SQL parser
+  + (\*) SQL -> HIR in `sql`
+  + `sequence_*` methods in adapter
+  + (\*) LIR -> dataflow in `compute`
+
+The two bullets marked (\*) above are a tricky point in the interface. SQL will _always_ lowers to **unstable**/**hot** HIR; we will have methods to convert **unstable**/**hot** HIR to the other two versions. Similarly, LIR lowers to dataflow from **unstable**/**hot** LIR; we will have methods to convert the other two versions to the latest.
+
+### Supporting qualification
+
+A capture-and-replay mode for events that stores the events in the customer's own environment would allow for (a) customers to easily test and qualify changes, and (b) a push-button way for us to spin up unbilled clusters to evaluate these replays in read-only mode.
```

Comment:
(a) would be nice, but would be a significant amount of work to add this feature.

An alternative way for customers to qualify the new optimizer version is to simply do a blue-green deployment where they change only the optimizer version, but not cut over to the new version for something like a week, and if it works fine over a week, then cut over. I think this would be minimal extra work for us.